### PR TITLE
[typeTag] Add new type tag parser for struct tags

### PR DESCRIPTION
**Pull request by gregnazario**
### Description
The typetag parser will need to be adjusted for generics and other special cases in an upcoming pull request.

This adds a much simpler parser and has comments for every choice made.

### Test Plan
I've put a pretty comprehensive set of tests, including some that are disabled that aren't properly handled in either parser (it can probably be updated in the new one).

Additionally, replaced structTag parsing with this implementation for now.

### CODE REVIEW

File: src/transactions/typeTag/typeTagParser.ts [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+
+/**
+ * All types are made of a few parts they're either:
+ * 1. A simple type e.g. u8
+ * 2. A standalone struct e.g. 0x1::account::Account
+ * 3. A nested struct e.g. 0x1::coin::Coin<0x1234::coin::MyCoin>
+ *
+ * There are a few more special cases that need to be handled, however.
+ * 1. Multiple generics e.g 0x1::pair::Pair<u8, u16>
+ * 2. Spacing in the generics e.g. 0x1::pair::Pair< u8 , u16>
```
Comment (xbtmatt) [inline src/transactions/typeTag/typeTagParser.ts:54]: Could you normalize all inputs by removing all whitespace and then re-adding it only after commas?

Comment (gregnazario) [reply src/transactions/typeTag/typeTagParser.ts:54]: Yeah, also normalization by removing spaces and then reparsing is dangerous, if you jam two words together when you forget to have commas.  (not like anyone should be naming their types like that)

File: src/transactions/typeTag/typeTagParser.ts [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+ * 3. Nested generics of different depths e.g. 0x1::pair::Pair<0x1::coin::Coin<0x1234::coin::MyCoin>, u8>
+ * 4. Generics for types in ABIs are filled in with placeholders e.g T1, T2, T3
+ */
+export function parseTypeTag(typeStr: string) {
+  const saved: Array<{ savedExpectedTypes: number; savedStr: string; savedTypes: Array<TypeTag> }> = [];
```
Comment (0xmaayan) [inline src/transactions/typeTag/typeTagParser.ts:59]: probably a better name
```suggestion
 type Saved = {
  savedExpectedTypes: number; 
  savedStr: string; 
  savedTypes: Array<TypeTag>
 }
  const saved: Array<Saved> = [];
```

File: src/transactions/typeTag/typeTagParser.ts [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+  // This represents the internal types for a type tag e.g. '0x1::coin::Coin<innerTypes>'
+  let innerTypes: Array<TypeTag> = [];
+  // This represents the current parsed types in a comma list e.g. 'u8, u8'
+  let curTypes: Array<TypeTag> = [];
+  // This represents the current character index
+  let cur = 0;
```
Comment (0xmaayan) [inline src/transactions/typeTag/typeTagParser.ts:65]: can we type each?
```
let cur: number = 0;
let currentStr: string = "";
let expectedTypes: number = 1;
```

File: src/transactions/typeTag/typeTagParser.ts [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+      // Process last type, if there is no type string, then don't parse it
+      if (currentStr !== "") {
+        const newType = parseTypeTagInner(currentStr, innerTypes);
+        curTypes.push(newType);
+      }
+
+      // Pop off stack outer type, if there's nothing left, there were too many '>'
+      const savedPop = saved.pop();
+      if (savedPop === undefined) {
+        throw new TypeTagParserError("Failed to parse typeTag, unexpected '>'");
```
Comment (0xmaayan) [inline src/transactions/typeTag/typeTagParser.ts:96]: can we log the typeTag?

File: src/transactions/typeTag/typeTagParser.ts [hunk 5]
```diff
@@ -0,0 +1,246 @@
+// Copyright © Aptos Foundation
```
Comment (0xmaayan) [inline src/transactions/typeTag/typeTagParser.ts:1]: I would rename the file to just `parser`, it is already under `typeTag` folder

File: src/transactions/typeTag/typeTagParser.ts [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+/* eslint-disable @typescript-eslint/no-unused-vars */
```
Comment (0xmaayan) [inline src/transactions/typeTag/typeTagParser.ts:4]: why?

File: src/transactions/typeTag/typeTagParser.ts [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+/* eslint-disable class-methods-use-this */
```
Comment (0xmaayan) [inline src/transactions/typeTag/typeTagParser.ts:5]: make them static?

File: src/transactions/typeTag/typeTagParser.ts [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+    } else if (char === ">") {
+      // Process last type, if there is no type string, then don't parse it
+      if (currentStr !== "") {
+        const newType = parseTypeTagInner(currentStr, innerTypes);
+        curTypes.push(newType);
+      }
+
+      // Pop off stack outer type, if there's nothing left, there were too many '>'
+      const savedPop = saved.pop();
+      if (savedPop === undefined) {
```
Comment (0xmaayan) [inline src/transactions/typeTag/typeTagParser.ts:95]: do we also need to check for `null`?

File: src/transactions/typeTag/typeTagParser.ts [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+        const newType = parseTypeTagInner(currentStr, innerTypes);
+
+        // parse type tag and push it on the types
+        innerTypes = [];
+        curTypes.push(newType);
+        currentStr = "";
+        expectedTypes += 1;
+      }
+    } else if (isValidWhitespaceCharacter(char)) {
+      // This means we should save what we have and everything else should skip until the next ,
```
Comment (0xmaayan) [inline src/transactions/typeTag/typeTagParser.ts:123]: nit
```suggestion
      // This means we should save what we have and everything else should skip until the next
```

File: src/transactions/typeTag/typeTagParser.ts [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+        curTypes.push(newType);
+        currentStr = "";
+        parsedTypeTag = true;
+      }
+
+      // Skip ahead on any more whitespace
+      cur = consumeWhitespace(typeStr, cur);
+
+      // The next space MUST be a comma, or a closing > if there was something parsed before
+      // e.g. `u8 u8` is invalid but `u8, u8` is valid
```
Comment (0xmaayan) [inline src/transactions/typeTag/typeTagParser.ts:139]: nice

File: src/transactions/typeTag/typeTagParser.ts [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+/**
+ * Parses a type tag with internal types associated
+ * @param str
+ * @param types
+ */
+function parseTypeTagInner(str: string, types: Array<TypeTag>): TypeTag {
+  switch (str) {
+    case "bool":
+      if (types.length > 0) {
+        throw new TypeTagParserError("Invalid typeTag, bool should not have a type argument");
```
Comment (0xmaayan) [inline src/transactions/typeTag/typeTagParser.ts:177]: not sure that I understand the error message..?

File: tests/unit/typeTagParser.test.ts [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+const MODULE_NAME = new Identifier("tag");
+const STRUCT_NAME = new Identifier("Tag");
+
+const structTagType = (typeTags?: Array<TypeTag>) =>
+  new TypeTagStruct(new StructTag(AccountAddress.ONE, MODULE_NAME, STRUCT_NAME, typeTags ?? []));
+
+describe("TypeTagParser", () => {
+  test("invalid types", () => {
+    // Missing 8
+    expect(() => parseTypeTag("8")).toThrow();
```
Comment (0xmaayan) [inline tests/unit/typeTagParser.test.ts:34]: can also add the expected message it throws to the test
```
toThrow("Missing 8")
```

**Review by 0xmaayan — APPROVED** [review]
Looks great! left some comments

File: src/transactions/typeTag/typeTagParser.ts (hunk 11 above)
Comment (gregnazario) [reply src/transactions/typeTag/typeTagParser.ts:177]: Maybe we make a better name but basically:

This is not a type
```
bool<u8>
```
This is the only type that's a bool
```
bool
```

File: src/transactions/typeTag/typeTagParser.ts (hunk 8 above)
Comment (gregnazario) [reply src/transactions/typeTag/typeTagParser.ts:95]: `pop` returns `undefined` or `value` as far as I could see.

File: src/transactions/typeTag/typeTagParser.ts (hunk 7 above)
Comment (gregnazario) [reply src/transactions/typeTag/typeTagParser.ts:5]: I should probably remove these, I believe this was copied in

File: src/transactions/typeTag/typeTagParser.ts (hunk 3 above)
Comment (gregnazario) [reply src/transactions/typeTag/typeTagParser.ts:65]: sure

File: tests/unit/typeTagParser.test.ts (hunk 12 above)
Comment (gregnazario) [reply tests/unit/typeTagParser.test.ts:34]: The messages need to be standardized a bit (which is why I didn't add any of them).

Will take a look at adding them.

**Pull request merged**